enterprise/server/ociregistry/ociregistry.go

```diff
@@ -186,29 +191,36 @@ func getImage(ctx context.Context, ref name.Reference) (v1.Image, error) {
 	return img, nil
 }
 
-// handleBlobRequest fetches all or part of a blob from a remote registry.
-// Used to pull OCI image layers.
-func (r *registry) handleBlobRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, imageName, refName string) {
-	reqStartTime := time.Now()
+func fetchLayerFromRemoteRegistry(imageName, layerSHA256Digest string) (gcr.Layer, error) {
+	d, err := name.NewDigest(imageName + "@sha256:" + layerSHA256Digest)
+	if err != nil {
+		return nil, fmt.Errorf("invalid hash %q: %s", layerSHA256Digest, err)
+	}
 
-	d, err := name.NewDigest(imageName + "@" + refName)
+	layer, err := remote.Layer(d)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("invalid hash %q: %s", refName, err), http.StatusNotFound)
-		return
+		return nil, fmt.Errorf("error fetching layer %q: %s", d, err)
 	}
+	return layer, nil
+}
 
-	layer, err := r.getBlob(ctx, d)
+// handleBlobRequest fetches all or part of a blob from a remote registry.
+// Used to pull OCI image layers.
+func (r *registry) handleBlobRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, imageName, blobSHA256Digest string) {
+	reqStartTime := time.Now()
+
+	layer, err := fetchLayerFromRemoteRegistry(imageName, blobSHA256Digest)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("error fetching layer %q: %s", d, err), http.StatusNotFound)
+		http.Error(w, err.Error(), http.StatusNotFound)
```

Comment:
since we're fetching from the remote here, would it be easy to preserve the original HTTP code from the remote error? I'm not sure if clients treat different codes differently, e.g. some codes might be retryable and we might be losing out on the opportunity to retry by returning 404

Reply:
I think the easiest way to do that is to make an HTTP request ourselves instead of relying on the `go-containerregistry` library. Happy to do that, but as part of a separate change!